Source/Core/VideoCommon/VertexCache.cpp

```diff
@@ -0,0 +1,257 @@
+// (though it doesn't explain how those work with the perf counter)
+
+constexpr u32 NUM_CACHE_SETS = 8;
+constexpr u32 NUM_CACHE_LINES = 8;
+constexpr u32 CACHE_LINE_SIZE = CACHE_RAM_SIZE / NUM_CACHE_SETS / NUM_CACHE_LINES;
+
+constexpr u32 CACHE_LINE_OFFSET_BITS = 7;
+static_assert(CACHE_LINE_SIZE == 1 << CACHE_LINE_OFFSET_BITS);
+constexpr u32 CACHE_LINE_OFFSET_SHIFT = 0;
+constexpr u32 CACHE_LINE_OFFSET_MASK = (1 << CACHE_LINE_OFFSET_BITS) - 1;
+
+constexpr u32 SET_NUMBER_BITS = 3;
+static_assert(NUM_CACHE_SETS == 1 << SET_NUMBER_BITS);
+constexpr u32 SET_NUMBER_SHIFT = CACHE_LINE_OFFSET_SHIFT + CACHE_LINE_OFFSET_BITS;
+constexpr u32 SET_NUMBER_MASK = (1 << SET_NUMBER_BITS) - 1;
+
+constexpr u32 TAG_BITS = 32 - CACHE_LINE_OFFSET_BITS - SET_NUMBER_BITS;
+constexpr u32 TAG_SHIFT = SET_NUMBER_SHIFT + SET_NUMBER_BITS;
+constexpr u32 TAG_MASK = (1 << TAG_BITS) - 1;
+
+// This is a guess, but the size should be proportional to memory latency
+constexpr u32 REQUEST_QUEUE_SIZE = 20;
+
+// The patent describes (Figure 5, pp. 9-10 lines 63-5) the format of a memory address, but without
+// giving the number of bits.  It says a memory address 450 is passed from the stream parser 208 to
+// the vertex cache 212, and is composed of:
+//
+// - a byte offset into the cache line (452)
+// - a tag RAM address (454)
+// - a main memory address (456, but not noted in the most obvious place), which is used for
+//   comparison with the tag RAM contents (the figure makes this a bit confusing, as the English
+//   figure names it as "Compared with Tag RAM Contents", but figure 10 (図１０) in the Japanese
+//   patent labels it explictly as "ﾒｲﾝﾒﾓﾘｱﾄﾞﾚｽ" (literally main memory address))
+//
+// Earlier text says that the vertex cache can be addressed as if it were the entirety of main
+// memory (p. 7 lines 51-53), so this is just the way a memory address is split up when being
+// interpreted by the vertex cache.
+//
+// For our purposes we can assign clearer names; instead of "tag RAM address" the middle field is
+// "set number" (s), and instead of "main memory address" the last field is "tag" (t).  The byte
+// offset is a clear enough name, and will be designated with o.
+// Thus, for 7 bits of cache line offset data and 3 bits for set number, an address is like this:
+// tttt tttt tttt tttt tttt ttss sooo oooo
+// Of course, on a real console there isn't enough RAM for all of the bits in the tag to matter,
+// but for ease of implementation, we'll use all of the bits, and also include the set bits,
+// since all RAM has the same speed and this will make debugging slightly easier.
+// (we can include the set bits since within a given set, the set bits will always be the same,
+// so all tags in the given set will have the same set bits)
+namespace
+{
+u32 GetCacheLineOffset(u32 address)
+{
+  return (address >> CACHE_LINE_OFFSET_SHIFT) & CACHE_LINE_OFFSET_MASK;
+}
+
+u32 GetSetNumber(u32 address)
+{
+  return (address >> SET_NUMBER_SHIFT) & SET_NUMBER_SHIFT;
```

Comment:
Should the latter SHIFT be MASK?

Reply:
Yes, but that ends up not mattering here because both SHIFT and MASK are 7 in this case (but that does cause problems if the cache line size is changed).  (I actually caught this yesterday but hadn't pushed a fix yet).